### PR TITLE
Build on Windows using rwinlib builds, closes #5

### LINKED ISSUE #5: netcdf4 and opendap support on Windows
Michael Sumner reports that RNetCDF on Windows is built with netcdf3. It would be useful to build with netcdf4 instead, so that netcdf4 classic files can read and modified. OPeNDAP support would also be useful.

### PR DESCRIPTION
**Pull request by jeroen**

### CODE REVIEW

**Comment by mjwoods** [comment]
Thanks @jeroen , that is extremely helpful of you.

I just tested your branch on both https://win-builder.r-project.org and https://builder.r-hub.io , and it installed perfectly.

Do you know if the netcdf library from rwinlib supports opendap?

File: src/Makevars.win [hunk 1]
```diff
@@ -1,6 +1,19 @@
-# Makevars.win can be tested by uploading package to win-builder.r-project.org
+VERSION=4.4.1.1
+PKG_CPPFLAGS = -I../windows/netcdf-${VERSION}/include \
+	-DHAVE_LIBUDUNITS2 -DHAVE_DECL_NC_RENAME_GRP=1
 
-PKG_CPPFLAGS = -DHAVE_LIBUDUNITS2 -DHAVE_DECL_NC_RENAME_GRP=1 \
-  -Drestrict=restrict
```
Comment (mjwoods) [inline src/Makevars.win]: I was using `restrict` deliberately, and it worked in my previous testing. Is there a reason to remove it?

Comment (jeroen) [inline src/Makevars.win]: Does this flag do anything?

**Comment by jeroen** [comment]
I haven't build it with opendap. Is that a feature in netcdf?

**Comment by jeroen** [comment]
It looks like currently it is not possible to build opendap on windows: https://github.com/OPENDAP/libdap4/issues/33. I might try to work with the opendap folks to get this fixed but for now we have to do without.

File: src/Makevars.win (hunk 1 above)
Comment (mjwoods) [inline src/Makevars.win]: It mainly serves as a reminder that the `restrict` keyword is being used, and that a different keyword may be needed by some compilers. This is useful on systems with older C compilers, but I suppose we don't care about them when building for Windows.

**Comment by mjwoods** [comment]
Hi @jeroen, opendap is a nice feature of netcdf. It lets you process a subset of a large file without transferring the whole file over a network. For example, you can display the header of a dataset with just a URL:

```shell
ncdump -h http://remotetest.unidata.ucar.edu/dts/test.01
```

I'm not sure that we actually need libdap. I managed to build netcdf on Linux without having libdap installed, and the resulting library allowed me to use opendap (i.e. the above `ncdump` command worked). I suspect that the netcdf4 source package includes some bundled opendap code.

Looking at the `configure.ac` file from netcdf (version 4.4.1.1), it seems that opendap is only enabled if libcurl is detected. So maybe adding libcurl to the rwinlib installation will enable opendap in netcdf. Would you mind giving that a try?

**Comment by mjwoods** [comment]
I have another question for you @jeroen . Are you sure that RNetCDF will build correctly on CRAN if it depends on rwinlib?

**Comment by jeroen** [comment]
What function in RNetCDF can I use to test if dap is working?

**Comment by jeroen** [comment]
Yes it will build on CRAN. The "winbuilder" website is the same build machine as cran; if it works there then it will build on CRAN. There are [more than 30 packages on cran](https://github.com/search?q=org%3Acran+rwinlib&type=Code) using rwinlib already.

**Comment by mjwoods** [comment]
Here is an example provided by Michael Sumner:

```r
u <- "http://coastwatch.pfeg.noaa.gov/erddap/griddap/erdQSwind3day"
library(RNetCDF)
var.get.nc(open.nc(u), "altitude")
[1] 10
```

If it doesn't work, you will see an error like "NetCDF: Unknown file format".

**Comment by jeroen** [comment]
The example just gives me **Error: NetCDF: Write to read only** on all platforms.

**Comment by jeroen** [comment]
Hmz with the 1.9 branch i can get it to work on win64 but on 32 bit it just crashes :/

**Comment by jeroen** [comment]
I have updated the builds and this PR to enable DAP. With this, the example seems to work on the 1.9 branch at least on Win64. There is still an issue on Windows 32 bit but this might be a bug in your package.

**Comment by mjwoods** [comment]
Thanks very much @jeroen .

I will need to investigate the crash on 32-bit platforms. Perhaps I have introduced a bug recently, because it was working on 32-bit Windows not long ago.

**Comment by jeroen** [comment]
The problem only happens for the DAP example. It might be that there is a problem in netcdf DAP on 32bit. All current functionality still work. I think you can merge this PR.

**Comment by mjwoods** [comment]
Hi @jeroen, the crash problem seems to be isolated to Win32 platforms. The DAP example worked perfectly for me on 32-bit Linux (CentOS6).

I notice that base R depends on libcurl. Are you using the same libcurl and compiler flags when building netcdf?

Do you have any instructions about building netcdf for rwinlib? What about build logs? If you could send them to me by email, they may help me to diagnose the problem.

**Comment by jeroen** [comment]
I'm building with [msys2](https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-netcdf/PKGBUILD) and rtools using cmake:

```
  ${MINGW_PREFIX}/bin/cmake -G "MSYS Makefiles" \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=ON \
    -DENABLE_DAP=ON \
    -DBUILD_TESTING=OFF \
    -DENABLE_DYNAMIC_LOADING=OFF \
    -DENABLE_NETCDF_4=ON \
    "${srcdir}/${_realname}-${pkgver}"
  make
```

**Comment by mjwoods** [comment]
Thanks for your help, @jeroen. I will close this merge request, because you have solved the important problem of building RNetCDF with netcdf4 support on Windows. As a bonus, the opendap support seems to be working on 64-bit Windows. I will raise a separate issue dealing with the crash problem when attempting to use opendap on 32-bit Windows.

**Comment by jeroen** [comment]
Perhaps we should just ship the build without DAP support for win32 if we can't get it to work at all.

**Comment by mjwoods** [comment]
I'd be happy with that solution. Do you know if `ncdump` works with DAP datasets on Win32?

**Comment by mjwoods** [comment]
Hi @jeroen , you suggested rebuilding netcdf for win32 without DAP support. I think that would be a good idea. Can you please do that for me on rwinlib? (That would also resolve https://github.com/rwinlib/netcdf/issues/1)

**Comment by jeroen** [comment]
OK I have fixed it in https://github.com/rwinlib/netcdf/commit/c630bc0fcbaf64a34581f1d882aa0f918c5b92c2 and updated the tag. 

There is nothing you need to do on your end. The next time you build RNetCDF on windows it will use the DAP version on win64 and the non DAP version on win32.

**Comment by mjwoods** [comment]
Thanks very much @jeroen !

**Comment by jeroen** [comment]
We have some fresh builds of netcdf and hdf5. Can you help test if the 32 bit crash is resolved?

**Comment by mdsumner** [comment]
@jeroen I'm able to test this, I see the updates at https://github.com/rwinlib/hdf5 but nothing at https://github.com/rwinlib/netcdf ?

**Comment by jeroen** [comment]
I already tried it myself and it doesn't work. It freezes as soon as you pass a url to `open.nc`. I don't know what these netcdf folks are testing with but I guess it isn't Windows.

**Comment by mjwoods** [comment]
Hi @jeroen , I found the same as @mdsumner - a new hdf5-1.10.5 but the old netcdf-4.4.1.1+dap. Is there supposed to be a newer netcdf release on rwinlib? For example, the latest stable release is netcdf-4.7.1.

I suspect that netcdf needs to be configured for the version of hdf5 it is linked against. I recall having problems linking hdf5-1.10.x with netcdf built for hdf5-1.8.x, and it may have been due to a change in the default integer type used for some hdf5 function interfaces. The netcdf `configure` script should be able to detect and adapt for changes like that.

**Comment by pmjherman** [comment]
On my github I have a version of the rncdf4 package that does work with openDAP on windows.

I am trying to get it working with the newest netcdf, although the version I have is not much behind. It is extremely tricky to get it compiled correctly. It appears as if something in curl is not working properly, but actually this is not the case.
I have been debugging netcdf using visual code. The weird thing is, when I compile the debug version it works with dap and all, but when I compile the production version it doensn't. I am chasing where the problem comes from. At present, I believe it has to do with memory allocation and not including malloc.h at all due places. But I am not sure. I have had to adapt the netcdf code in many places for it to compile correctly. It has been some time since I last took this up, but I will restart soon and adjust my github site if I am successful

**Comment by jeroen** [comment]
@pmjherman we have automated the patching and build in https://github.com/r-windows/rtools-packages. You can easily kick off a build by sending a PR to there. You can also easily build yourself using rtools40 compilers.

**Pull request merged**